File: dsa-js/arrays/contains-duplicate.js
/**
 * <h2>Contains Duplicate</h2>
 * @see {@link https://neetcode.io/solutions/contains-duplicate}
 */

/**
 * <h3>Approach 1: Sorting</h3>
 * <p>Sorts the array and checks if adjacent elements are identical.</p>
 * <ul>
 * <li><b>Time Complexity:</b> O(n log n)</li>
 * <li><b>Space Complexity:</b> O(1) or O(n) depending on JS engine sort implementation</li>
 * </ul>
 */
function hasDuplicateSort(nums) {
    // JS sort is lexicographical by default; (a, b) => a - b is required for numbers
    nums.sort((a, b) => a - b);
    for (let i = 1; i < nums.length; i++) {
        if (nums[i] === nums[i - 1]) {
            return true;
        }
    }
    return false;
}

/**
 * <h3>Approach 2: Hash Set</h3>
 * <p>Uses a Set to track seen numbers for O(1) average lookup time.</p>
 * <ul>
 * <li><b>Time Complexity:</b> O(n)</li>
 * <li><b>Space Complexity:</b> O(n)</li>
 * </ul>
 */
function hasDuplicateSet(nums) {
    const seen = new Set();
    for (const num of nums) {
        if (seen.has(num)) {
            return true;
        }
        seen.add(num);
    }
    return false;
}

module.exports = {
    hasDuplicateSort,
    hasDuplicateSet
};

File: dsa-java/app/src/main/java/com/nik/dsa/neetcode250/arrays/ContainsDuplicate.java
package com.nik.dsa.neetcode250.arrays;

import java.util.HashSet;
import java.util.Set;

/**
 * <h2>LeetCode 217: Contains Duplicate</h2>
 * <p>Given an integer array nums, return true if any value appears at least twice
 * in the array, and return false if every element is distinct.</p>
 * @see <a href="https://leetcode.com/problems/contains-duplicate/">Problem Link</a>
 */
public class ContainsDuplicate {

    /**
     * <h3>Hash Set Approach</h3>
     * <p>Uses a HashSet to store visited elements. If an element is already present in the set,
     * a duplicate has been found.</p>
     * <ul>
     * <li><b>Time Complexity:</b> O(n) - We iterate through the array once; set lookups are O(1) average.</li>
     * <li><b>Space Complexity:</b> O(n) - In the worst case (no duplicates), all elements are stored in the set.</li>
     * </ul>
     */
    public boolean hasDuplicate(int[] nums) {
        Set<Integer> seen = new HashSet<>();
        for (int num : nums) {
            if (seen.contains(num)) {
                return true;
            }
            seen.add(num);
        }
        return false;
    }
}
